### PR TITLE
Progress reporting improvements

### PR DESCRIPTION
**Pull request by Ralith**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

**Review by martinvonz — COMMENTED** [review]
It looks like most of the changes in this PR are things already implemented in the `indicatif` crate. What can we not get from that crate?

**Comment by Ralith** [comment]
I initially attempted to use indicatif. I hit multiple issues:
- No built-in support for delayed initial visibility
- `ProgressBar` wants to be passed the total size of the operation (e.g. byte count) on construction. `libgit2` does not expose this operation. We could pass object+delta count or some other arbitrary large value, but the provided transfer rate display would produce garbage. We could implement custom transfer rate feedback using [`ProgressStyle::with_key`](https://docs.rs/indicatif/latest/indicatif/style/struct.ProgressStyle.html#method.with_key), but this can only be done through a `'static` `Fn`, making it awkward to access the actual rate data we're trying to display.
- The library appears to demand exclusive control of the terminal, making it impossible to mitigate the above API issues by including our own output alongside the bar.
- The library wants to output *directly* to the terminal, which is at odds with jj's routing output through `Ui`. This can be overridden only by implementing [`TermLike`](https://docs.rs/indicatif/latest/indicatif/trait.TermLike.html), which requires reimplementing some of the very logic we were reaching for an external lib to provide in the first place.

For those reasons, it seems like a satisfactory implementation on indicatif would require enough boilerplate and hacks to somewaht undermine the benefit of a library. Meanwhile, it adds several thousand lines of code in itself and a nontrivial dependency closure, while the implementation here is less than two hundred LoC, plus crossterm.

File: src/progress.rs [hunk 1]
```diff
@@ -62,6 +69,23 @@ impl Drop for Progress<'_> {
     }
 }
 
+fn draw_progress(progress: f32, buffer: &mut String, width: usize) {
+    const CHARS: [char; 9] = [' ', '▏', '▎', '▍', '▌', '▋', '▊', '▉', '█'];
```
Comment (martinvonz) [inline src/progress.rs:68]: The last one seems to be unused, right?

**Review by martinvonz — APPROVED** [review]

**Comment by martinvonz** [comment]
> I initially attempted to use indicatif. I hit multiple issues:
> 
> * No built-in support for delayed initial visibility
> * `ProgressBar` wants to be passed the total size of the operation (e.g. byte count) on construction. `libgit2` does not expose this information. We could pass object+delta count or some other arbitrary large value, but then the provided transfer rate display would produce garbage. We could implement custom transfer rate feedback using [`ProgressStyle::with_key`](https://docs.rs/indicatif/latest/indicatif/style/struct.ProgressStyle.html#method.with_key), but this can only be done through a `'static` `Fn`, making it awkward to access the actual rate data we're trying to display.
> * The library appears to demand exclusive control of the terminal, making it impossible to mitigate the above API issues by including our own output alongside the bar.
> * The library wants to output _directly_ to the terminal, which is at odds with jj's routing output through `Ui`. This can be overridden only by implementing [`TermLike`](https://docs.rs/indicatif/latest/indicatif/trait.TermLike.html), which requires reimplementing some of the very logic we were reaching for an external lib to provide in the first place.
> 
> For those reasons, a satisfactory implementation might require enough boilerplate and hacks to somewaht undermine the benefit of a library. Meanwhile, it adds several thousand lines of code in itself and a nontrivial dependency closure, while the implementation here is less than two hundred LoC, plus crossterm.

Fair enough. Thanks for explaining.

File: src/progress.rs [hunk 2]
```diff
@@ -62,6 +69,23 @@ impl Drop for Progress<'_> {
...
+    const RESOLUTION: usize = CHARS.len() - 1;
+    let ticks = (width as f32 * progress * RESOLUTION as f32).round() as usize;
```
[the remark this answers is not in the document]
Comment (Ralith) [reply src/progress.rs]: Done.

File: src/progress.rs [hunk 3]
```diff
@@ -62,6 +69,23 @@ impl Drop for Progress<'_> {
...
+    let whole = ticks / RESOLUTION;
+    for _ in 0..whole {
+        buffer.push('█');
+    }
+    if whole < width {
+        let fraction = ticks - whole * RESOLUTION;
```
[the remark this answers is not in the document]
Comment (Ralith) [reply src/progress.rs]: Good call.

File: src/progress.rs (hunk 1 above)
Comment (Ralith) [reply src/progress.rs:68]: Refactored to pull everything from this array.

**Pull request merged**